Introducing UDP listener.

conqerAtapple opened the pull request:
Signed-off-by: Jojy G Varghese <jojy_varghese@apple.com>

For an explanation of how to fill out the fields, please see the relevant section 
in [PULL_REQUESTS.md](./PULL_REQUESTS.md)

*Description*: Base layer for UDP proxying. UDP listener provides a `Buffer` interface to the upper layers, which in turn can add features/services like filters on top.
*Risk Level*: Medium
*Testing*: Unit testing
*Docs Changes*: n/a
*Release Notes*: n/a
[Optional Fixes #Issue]
[Optional *Deprecated*:]

zyfjeff commented on source/common/network/udp_listener_impl.cc:
```diff
@@ -0,0 +1,119 @@
+void UdpListenerImpl::enable() { event_add(&raw_event_, nullptr); }
+
+void UdpListenerImpl::readCallback(int fd, short flags, void* arg) {
+  (void)flags;
```
UNREFERENCED_PARAMETER?

conqerAtapple commented:
@mattklein123 @alyssawilk @mpwarres  PTAL

This PR is first of the series for UDP proxying. This is extracted from the WIP #5256. I will eventuallu close #5246.

mpwarres commented on include/envoy/event/dispatcher.h:
```diff
@@ -110,6 +110,16 @@ class Dispatcher {
+   * @param bind_to_port controls whether the listener binds to a transport port or not.
+   * @return Network::ListenerPtr a new listener that is owned by the caller.
+   */
+  virtual Network::ListenerPtr createUdpListener(Network::Socket& socket,
```
Wondering whether we will ever want to support proxying datagrams over Unix domain sockets, in addition to over UDP. (Analogous to current Envoy support for pipes vs. TCP connections.) If so, then it might be more general to name this createDatagramListener. Or maybe UDS datagrams aren't worth supporting, in which case current name SGTM.

mpwarres replied:
Not that I'm aware. We have used SOCK_DGRAM + shared memory to allow GFEs (speaking QUIC) to efficiently front for other (non QUIC speaking) servers, but AIUI there the datagrams were just for coordinating shared memory use, not for carrying QUIC traffic. Not sure if @ianswett may have mentioned this in one of his talks, perhaps [this one](https://netdevconf.org/0x12/session.html?developing-and-deploying-a-tcp-replacement-for-the-web), though I don't see it in the slides.  The comment I posted upthread was more hypothetical--I don't have a concrete use case.

mattklein123 replied:
I don't have a strong opinion on this, though `createDatagramListener()` does seem more generic.

mpwarres commented on include/envoy/network/listener.h:
```diff
@@ -114,6 +114,46 @@ class ListenerCallbacks {
+   * @param peer_address Network address of the peer.
+   * @param data Data buffer received.
+   */
+  virtual void onData(Address::InstanceConstSharedPtr local_address,
```
Relaying a comment originally made by ianswett@chromium.org on an earlier listener sketch, which also applies here: from the perspective of QUIC (which will eventually use this), it would be desirable to bundle (local_address, peer_address, data) in a Datagram or Packet object, to accomodate possible future addition of other attributes, e.g. TTL. E.g., something like [this](https://gist.github.com/mpwarres/188e71dce411577d63505bde49b4a71c).

ianswett replied:
+1

conqerAtapple replied:
@mpwarres , @ianswett  Do you think we should have it here or in the QUIC filter layer? Do other UDP filters need the `Datagram`/`Packet` concept?

mattklein123 replied:
Seems fine to just make a struct now to carry the data if we want that? That will avoid a lot of code site changes later to add new fields.

alyssawilk replied:
+1, let's totally make it a struct sooner rather than later.

mpwarres commented on include/envoy/network/listener.h:
```diff
@@ -114,6 +114,46 @@ class ListenerCallbacks {
+   */
+  virtual void onError(const ErrorCode& err_code, int err_no) PURE;
+};
+
```
Seems like the listener might also need to be notified when the UDP socket is writable? AIUI, in the TCP case, this interaction is a little hidden: ListenerCallbacks::onNewConnection() provides a Network::Connection, and that Connection is set up with the Dispatcher to receive events when the connection socket is readable or writable: [link](https://github.com/envoyproxy/envoy/blob/f8ba63a4b2a3f0293cfcb44eddcab178f5adc059/source/common/network/connection_impl.cc#L65).  The "is readable" transition is already covered by UdpListenerCallbacks::onData(), but seems like there will need to be something somewhere for the "is writable" side, to let the receipient know when it can flush/write more packets.

conqerAtapple replied:
Good point. So I guess we need another callback `onWriteReady` ?

mpwarres replied:
I think it depends on the UdpListenerCallbacks vs. ListenerCallbacks question. If you go with the already existing ListenerCallbacks interface like @alyssawilk is suggesting, then it would make sense to reflect writability inside the Connection implementation, i.e. Event::FileReadyType::Write -> UdpConnectionImpl::onWriteReady() which could flush any buffered packets that are waiting to be written, paired with a UdpConnectionImpl::write() implementation that buffers packets when the socket is not currently writable.

If you go with UdpListenerCallbacks, then yeah, I guess there would need to be something like an onWriteReady(), and it would need to supply something to write to (either the Socket, or an interface for writing a packet).

conqerAtapple replied:
Not sure how to reuse existing `ListenerCallbacks` for UDP since it has connection concepts. I am leaning towards having `onWriteReady` in `UdpListener`. Let me know if you (@alyssawilk) feel strongly about this.

mattklein123 replied:
This is related to my comment above about whether we are doing connection oriented or not. IMO I'm leaning towads splitting UDP into 2 types: 1) What you have now, which is geared towards datagram, and 2) One listener type which uses normal listener callbacks, and spits out connections which then supports existing filters. Thoughts?

alyssawilk commented on include/envoy/network/listener.h:
```diff
@@ -114,6 +114,46 @@ class ListenerCallbacks {
+/**
+ * UDP listener callbacks.
+ */
+class UdpListenerCallbacks {
```
I thought Matt requested on the prior PR that we try to stick to listenercallbacks where possible?

I'd think the current code could be a subclass of ListenerCallbacks, where onAccept was commeted as "accepting new data" rather than "accepting a new connection", we call onNewConnection where we currently do (but with a UdpConnection subclass which may be filled with TODOs, but we pass the local address and the peer address to it) and then hand onData calls to the new connection?

conqerAtapple replied:
Same comment as below. The current `ListenerCallbacks` interface exposes connection concept. Inheriting that interface would semantically mean - `UdpListenerCallbacks` `is-a` `ListenerCallbacks`. Not sure if we mean that. What about having a `bridge` that bridges the UDP listener to the existing `ListenerCallbacks`? This bridge could be a concrete implementation of a `UdpListenerCallbacks` interface.  Thinking out loud here. Would he happy to hear what you think.

mattklein123 replied:
See my comments below.

conqerAtapple commented on include/envoy/event/dispatcher.h:
```diff
@@ -110,6 +110,16 @@ class Dispatcher {
+   * Create a logical udp listener on a specific port.
+   * @param socket supplies the socket to listen on.
+   * @param cb supplies the udp listener callbacks to invoke for listener events.
+   * @param bind_to_port controls whether the listener binds to a transport port or not.
```
makes sense

conqerAtapple commented on source/common/buffer/buffer_impl.cc:
```diff
@@ -107,6 +107,32 @@ void OwnedImpl::move(Instance& rhs, uint64_t length) {
+Api::SysCallIntResult OwnedImpl::recvFrom(int fd, uint64_t max_length, sockaddr_storage& peer_addr,
+                                          socklen_t& addr_len) {
+  if (max_length == 0) {
+    return {0, 0};
```
I was trying to be consistent with the same case handled elsewhere in the code.

conqerAtapple replied:
Thinking about it a little more,  here the system call is not invoked so would it be reflective of the real state of the process to change `errno` ?

mpwarres commented on source/common/buffer/buffer_impl.cc:
```diff
@@ -107,6 +107,32 @@ void OwnedImpl::move(Instance& rhs, uint64_t length) {
+  const uint64_t num_slices = reserve(max_length, &slice, 1);
+
+  ASSERT(num_slices == 1);
+  // TODO(conqerAtapple): Use os_syscalls
```
The Chromium QUIC code can do either. QuicPacketReader::ReadAndDispatchPackets() reads packets either via [recvmsg](https://cs.chromium.org/chromium/src/net/third_party/quic/platform/impl/quic_socket_utils.cc?l=228&rcl=5553a1588f5fbf1bd7f6047102f1446f26ee3172) or [recvmmsg](https://cs.chromium.org/chromium/src/net/third_party/quic/core/quic_packet_reader.cc?l=94&rcl=5553a1588f5fbf1bd7f6047102f1446f26ee3172) depending on whether the macro MMSG_MORE is set (see [QuicPacketReader::ReadAndDispatchPackets](https://cs.chromium.org/chromium/src/net/third_party/quic/core/quic_packet_reader.cc?l=68&rcl=5553a1588f5fbf1bd7f6047102f1446f26ee3172)()). Then, independent of whether the packets were read individually or as a batch, they are passed one by one into the ProcessPacket() method of either QuicDispatcher (if server) or QuicClientEpollNetworkHelper (if client).

alyssawilk replied:
Yeah, I think both for QUIC and for other UDP applications we're going to want to read-many then dispatch-one-at-a-time.  While I don't think we want to implement that code right now (launch and iterate!) I just want to make sure we design the APIs for it.

Then again I'm also fine landing this as is and moving it later - I don't personally mind a lot of code churn while we feel our way out as long as we all know what we're getting into :-P

conqerAtapple replied:
I agree that we need a `recvmmsg` implementation also (thanks to @mpwarres for clarifying). I am not sure how do we pass configurable parameters (number of messages for recvmmsg) to the listener. At this this I am leaning towards adding TODO for recvmmsg and attack that in my next PR.

cmluciano commented on include/envoy/network/listener.h:
```diff
@@ -114,6 +114,46 @@ class ListenerCallbacks {
+
+  /**
+   * On first packet received on a UDP socket. This allows the callback handler
+   * to establish filter chain (or any other prepararion).
```
```suggestion
   * to establish filter chain (or any other preparation).
```

mattklein123 commented on include/envoy/buffer/buffer.h:
```diff
@@ -168,6 +168,9 @@ class Instance {
    */
   virtual Api::SysCallIntResult read(int fd, uint64_t max_length) PURE;
 
+  virtual Api::SysCallIntResult recvFrom(int fd, uint64_t max_length,
```
nit: doc comments. As a larger comment, I would love to remove any type of file IO from the buffer interface moving forward since in the long term I don't think it fits with our FD refactors. Any thoughts on how we might not add it for this case since it's new? cc @brian-pane who is working on a new buffer impl.

mattklein123 commented on include/envoy/network/listener.h:
```diff
@@ -114,6 +114,46 @@ class ListenerCallbacks {
+   * @param peer_address Network address of the peer.
+   * @param data Data buffer received.
+   */
+  virtual void onNewConnection(Address::InstanceConstSharedPtr local_address,
```
I assume for the purpose of this PR, we are looking at stateless vs. connection oriented proxy, right? I guess my question comes down to, at this layer, what does `onNewConnection()` give us. It seems like there must be internal tracking to figure out whether this is "new", but then how are connections ever going to get cleaned up? Where is the interface for that? It seems like at this datagram layer, we could just have `onData()` and these are the interfaces for raw datagram listening. If we want connection oriented, could we just use the normal ListenerCallbacks which spit out normal connections which are then UdpConnectionImpl? Thoughts on this?

conqerAtapple replied:
@mattklein123  
Thanks for reviewing.

What about layering the connection concept on top of UDP/datagram? This would allow us to have 1 common `buffer` layer and then connections (eg QUIC) on top of it. When a connection oriented filter (QUIC) gets the `onNewConnection` event, it can then decide to do things like 1) prepare the protocol layers 2) Carve a protocol specific `Packet` and send it up the protocol layer.

mattklein123 replied:
The problem is I don't understand how this layer is going to do conn tracking if the API looks like this. I think you need an API to "delete" a connection if that's the case?

mattklein123 replied:
But then I think the `onNewConnection` API needs to be removed? What does a new connection mean without conn tracking?

conqerAtapple replied:
Yes that is a good point. The book keeping of `if this is the first data packet` has to be done somewhere, assuming that the state of `first data` is important to track. If it is not, then the API `onNewConnection` need not be there.

mattklein123 replied:
Yeah IMO if we are going to build connection oriented on top of this, I would kill `onNewConnection` and any tracking from this PR.

conqerAtapple commented:
@mattklein123 @mpwarres @alyssawilk @cmluciano 

Thanks for reviewing. I am summarizing the changes:

- Will fix spellchecks/comments suggestions
- Will add `UdpPacket` for `onData` (and `onNewConnection`)
- Will add `onWriteReady`
- Move `recvFrom` from buffer

In my next PR(s):
- Add implementation for `recvmmsg`
- Flush out the layering (ListenerCallbacks , UdpListenerCallbacks)

What do you guys think?

alyssawilk commented on include/envoy/network/listener.h:
```diff
@@ -114,6 +114,51 @@ class ListenerCallbacks {
+   * @param error_code ErrorCode for the error event.
+   * @param errno System error number.
+   */
+  virtual void onError(const ErrorCode& err_code, int err_no) PURE;
```
super nitty, but can you use full variable names?
error_code error_number etc.?

alyssawilk replied:
I think there's still a few of these outstanding.  Mind fixing up?

conqerAtapple replied:
I fixed them (latest commits) but looking at your comment again, I missed fixing `err_no` as you suggested. Will fix that.

alyssawilk commented on include/envoy/event/dispatcher.h:
```diff
@@ -110,6 +110,14 @@ class Dispatcher {
+   * @param cb supplies the udp listener callbacks to invoke for listener events.
+   * @return Network::ListenerPtr a new listener that is owned by the caller.
+   */
+  virtual Network::ListenerPtr createDatagramListener(Network::Socket& socket,
```
I was about to ask that we be consistent with Udp vs Datagram when I realized you were totally doing this right per mpw@'s advice.
I think when we get to docs it's worth doccing up that this is meant to be generic (potentially extending to datagrams over sockets) where the listener is for UdpSockets for others who have the same question :-P

alyssawilk commented on test/common/network/listener_impl_test.cc:
```diff
@@ -171,6 +259,402 @@ TEST_P(ListenerImplTest, UseActualDst) {
+/**
+ * Tests UDP listener for actual destination and data.
+ */
+TEST_P(ListenerImplTest, UseActualDstUdp) {
```
now that we have udp_listener_impl.* files, we should probably move shared test code to a common utility class and put these tests in /test/common/network/udp_listener_impl_test, yeah?

conqerAtapple commented on include/envoy/network/listener.h:
```diff
@@ -114,6 +114,51 @@ class ListenerCallbacks {
+ */
+struct UdpData {
+  Address::InstanceConstSharedPtr local_address_;
+  Address::InstanceConstSharedPtr peer_address_;
```
good point . Will change it to unique_ptr

conqerAtapple replied:
@alyssawilk I was looking at it again and it looks like it might be difficult to change `local_address` to a unique_ptr as it is derived from the  listener's `Socket` object. So I am thinking we could just have the `peer_address` as unique_ptr. It will convey the right ownership semantics but will look odd. WDYT?

conqerAtapple replied:
Wanted to clarify that I was talking about `local_address` that is from the listening socket. `peer_address` could be unique for each packet as its derived from the UDP packet. I assume you mean the same but just wanted to be clear :)

conqerAtapple replied:
@alyssawilk I have updated comments based on ☝️

conqerAtapple commented:
@alyssawilk PR is ready for another review. PTAL.

conqerAtapple commented:
@mattklein123 @alyssawilk ping

alyssawilk commented:
Oh sorry, missed this yesterday.  I'm out this morning but I'll try to do a pass this afternoon and sorry for the delay

conqerAtapple commented:
@mpwarres @alyssawilk Thanks for the review. I think all the suggested changes are in now.

mattklein123 commented:
@conqerAtapple can you merge master and then I will take a final pass?

conqerAtapple commented:
@mattklein123 I have rebased with master. Looks like "release" test is failing with message "server shutdown abruptly".

mattklein123 commented:
I think we are running out of RAM in CI. We are working on that in parallel cc @lizan. 

/retest

mattklein123 commented on include/envoy/network/listener.h:
```diff
@@ -114,6 +114,59 @@ class ListenerCallbacks {
 
+/**
+ * Utility struct that encapsulates the information from a udp socket's
+ * recvfrom/recvmmdg call.
```
typo "recvmmdg"

mattklein123 commented on source/common/event/dispatcher_impl.cc:
```diff
@@ -119,6 +120,12 @@ DispatcherImpl::createListener(Network::Socket& socket, Network::ListenerCallbac
+Network::ListenerPtr DispatcherImpl::createUdpListener(Network::Socket& socket,
+                                                       Network::UdpListenerCallbacks& cb) {
+  ASSERT(isThreadSafe());
+  return Network::ListenerPtr{new Network::UdpListenerImpl(*this, socket, cb)};
```
nit: `std::make_unique<>`

mattklein123 commented on source/common/network/udp_listener_impl.cc:
```diff
@@ -0,0 +1,163 @@
+UdpListenerImpl::UdpListenerImpl(const Event::DispatcherImpl& dispatcher, Socket& socket,
+                                 UdpListenerCallbacks& cb)
+    : BaseListenerImpl(dispatcher, socket), cb_(cb) {
+  event_assign(&raw_event_, &dispatcher.base(), socket.ioHandle().fd(),
```
Any reason to not just use a FileEvent here?

conqerAtapple replied:
I could have used `FileEvent` but was not familiar with the API, especially how the enable/disable functionality would work.

mattklein123 replied:
You just call `setEnabled()`, which IIRC has no events specified will disable everything. See https://github.com/envoyproxy/envoy/blob/master/include/envoy/event/file_event.h#L47. It's a pretty simple interface/impl. I think we should use a FileEvent under the hood to avoid duplicating the logic here and more easily allow us to move away from libevent if needed in the future.

conqerAtapple replied:
I have updated the PR with the change suggested.

mattklein123 commented on source/common/network/udp_listener_impl.cc:
```diff
@@ -0,0 +1,163 @@
+}
+
+void UdpListenerImpl::handleReadCallback(int fd) {
+  RELEASE_ASSERT(fd == socket_.ioHandle().fd(),
```
I would use a normal `ASSERT` here. Should this ever happen?

mattklein123 commented on source/common/network/udp_listener_impl.cc:
```diff
@@ -0,0 +1,163 @@
+}
+
+void UdpListenerImpl::eventCallback(int fd, short flags, void* arg) {
+  RELEASE_ASSERT((flags & (EV_READ | EV_WRITE)),
```
I would use a normal `ASSERT` here. Should never happen.

mattklein123 commented on source/common/network/udp_listener_impl.cc:
```diff
@@ -0,0 +1,163 @@
+      return;
+    }
+
+  } while (recv_result.result_.rc_ == -EAGAIN);
```
I'm confused. Won't this spin? Shouldn't we return if we get EAGAIN and wait for a read event?

conqerAtapple replied:
Good catch.

mattklein123 commented on source/common/network/udp_listener_impl.cc:
```diff
@@ -0,0 +1,163 @@
+}
+
+void UdpListenerImpl::handleWriteCallback(int fd) {
+  RELEASE_ASSERT(fd == socket_.ioHandle().fd(),
```
normal `ASSERT` if should never happen.

conqerAtapple commented:
@mattklein123 Thanks for review. I have addressed all comments , except the FileEvent one as I am not very clear on how the enable/disable functionality would work.

mattklein123 commented on include/envoy/event/dispatcher.h:
```diff
@@ -90,7 +90,7 @@ class Dispatcher {
    */
   virtual FileEventPtr createFileEvent(int fd, FileReadyCb cb, FileTriggerType trigger,
-                                       uint32_t events) PURE;
+                                       uint32_t events) const PURE;
```
We are somewhat splitting hairs here, but IMO this is not a const operation on the dispatcher, it does modify its state. Instead of changing this to a const method can you pass a non-const Dispatcher to where you need it? This is what is done elsewhere.

conqerAtapple commented:
@mattklein123 I have addressed all comments now. I think coverage tests are failing because of memory issues.

mattklein123 commented:
/retest

mattklein123 commented on source/common/event/dispatcher_impl.h:
```diff
@@ -30,7 +30,7 @@ class DispatcherImpl : Logger::Loggable<Logger::Id::main>, public Dispatcher {
    * @return event_base& the libevent base.
    */
-  event_base& base() { return *base_; }
+  event_base& base() const { return *base_; }
```
Can you revert this const change also and a couple of the other ones you added?

conqerAtapple replied:
I will remove them but I think they should be const :)

conqerAtapple commented:
@mattklein123 The `release` check has once again failed with `server terminated abruptly`

mattklein123 commented:
@conqerAtapple we are working on the CI issues, in the meantime you can rerun failed tests yourself, like so:

/retest
